hmmlearn/_hmmc.pyx

```diff
@@ -92,34 +91,58 @@ def _compute_lneta(int n_samples, int n_components,
 
 
 def _viterbi(int n_samples, int n_components,
-        np.ndarray[dtype_t, ndim=1] log_startprob,
-        np.ndarray[dtype_t, ndim=2] log_transmat,
-        np.ndarray[dtype_t, ndim=2] framelogprob):
+        dtype_t[:] log_startprob,
+        dtype_t[:, :] log_transmat,
+        dtype_t[:, :] framelogprob):
 
-    cdef int t, max_pos
-    cdef np.ndarray[dtype_t, ndim = 2] viterbi_lattice
-    cdef np.ndarray[np.int_t, ndim = 1] state_sequence
+    cdef int c0, c1, t, max_pos
+    cdef dtype_t[:, ::1] viterbi_lattice
+    cdef int[::1] state_sequence
     cdef dtype_t logprob
-    cdef np.ndarray[dtype_t, ndim = 2] work_buffer
+    cdef dtype_t[:, ::1] work_buffer
+    cdef dtype_t buf, maxbuf
 
     # Initialization
-    state_sequence = np.empty(n_samples, dtype=np.int)
+    state_sequence_arr = np.empty(n_samples, dtype=np.int32)
+    state_sequence = state_sequence_arr
     viterbi_lattice = np.zeros((n_samples, n_components))
-    viterbi_lattice[0] = log_startprob + framelogprob[0]
-
-    # Induction
-    for t in range(1, n_samples):
-        work_buffer = viterbi_lattice[t-1] + log_transmat.T
-        viterbi_lattice[t] = np.max(work_buffer, axis=1) + framelogprob[t]
+    work_buffer = np.empty((n_components, n_components))
 
-    # Observation traceback
-    max_pos = np.argmax(viterbi_lattice[n_samples - 1, :])
-    state_sequence[n_samples - 1] = max_pos
-    logprob = viterbi_lattice[n_samples - 1, max_pos]
+    with nogil:
+        for c1 in range(n_components):
+            viterbi_lattice[0, c1] = log_startprob[c1] + framelogprob[0, c1]
 
-    for t in range(n_samples - 2, -1, -1):
-        max_pos = np.argmax(viterbi_lattice[t, :] \
-                + log_transmat[:, state_sequence[t + 1]])
-        state_sequence[t] = max_pos
+        # Induction
+        for t in range(1, n_samples):
+            for c0 in range(n_components):
```

Comment:
We use `i` and `j` for states throughout the codebase. Sorry for being a bit pedantic :)